### PR TITLE
fix for v9 SON files

### LINKED ISSUE #818: SpikeIO does not load .smr
Dear Neo developers,
I can not load some of my .smr files with Spike2IO. I work with **python 3.5.2** and tried Neo versions **0.8.0** and **0.9.0.dev**. The error message is: “_IndexError: index 0 is out of bounds for axis 0 with size 0_”.
Note that I got the same error for some other  .smr files and could resolve this by using try_signal_grouping=False, however, the issue persists for some other .smr files. 
You can download an example file here: https://www.dropbox.com/s/hqgs4biuf5bm237/EP147-148-150-154_44d000.smr?dl=0
This file contains several channels (e.g. 'HCi1') of LFP recordings.

The code to reproduce the error:

from neo.io import Spike2IO
r = Spike2IO(filename='myfilename.smr',try_signal_grouping=False)

Thank you very much for your support.

### PR DESCRIPTION
**Pull request by MarinManuel**
In the CED documentation for the SON library, it says:

> In version 9 and above files, if pos is of type TDOF, the disk offset is pos*DISKBLOCK. In version 8 and previous files, the disk offset is pos. 

this PR fixes the offsets for v9 files, and fixes #818

### CODE REVIEW

**Comment by samuelgarcia** [comment]
Hi @MarinManuel. This is super cool and should help many users.
Thanks a lot.

**Comment by samuelgarcia** [comment]
I have push the file "Two-mice-bigfile-test000.smr" 
See https://gin.g-node.org/NeuralEnsemble/ephy_testing_data/src/master/spike2

Could add it [here in neo/test/rawiotest](https://github.com/NeuralEnsemble/python-neo/blob/master/neo/test/rawiotest/test_spike2rawio.py#L16) and [here in neo/test/iotest](https://github.com/NeuralEnsemble/python-neo/blob/master/neo/test/iotest/test_spike2io.py#L21) and check that test are passing ?

**Comment by JuliaSprenger** [comment]
The tests are failing as `channel_names` are expected to be of type 'S' and not unicode. @samuelgarcia was there a reason for this not to be unicode?
There are two options now:
1) change the definition of the `ChannelIndex` class and make `channel_name` unicode. This affects  all ios based on the raw mechanism.
2) Hackily reading the `channel_names` as unicode and converting it to type `S`.

I would go for approach 2, since the `ChannelIndex` object will anyway be replaced in the future (#817), so we don't need to introduce unnecessary changes for `ChannelIndex` objects now.

**Comment by MarinManuel** [comment]
@JuliaSprenger @samuelgarcia  I've added a line to convert back the unicode strings to ascii for the time being. Hopefully that does not introduce any more problems.

**Comment by samuelgarcia** [comment]
Hi all.
@JuliaSprenger this is strange. I thought we fix this old (and very anoying) behavior where everything was ascii string because of python 2. Now python27 is no more here and evrything should be unicode everwhere.

@MarinManuel : I propose to fix this str vs unicode problem outside this PR very very soon. So you could rebase this PR with the offcial fix. is it OK ?

**Comment by JuliaSprenger** [comment]
@MarinManuel the format change is in master now, could you rebase and undo your changes in `basefromrawio.py`?

File: neo/core/channelindex.py [hunk 1]
```diff
@@ -178,7 +178,7 @@ def __init__(self, index, channel_names=None, channel_ids=None,
             channel_ids = np.array([], dtype='i')
 
         # Store recommended attributes
-        self.channel_names = np.array(channel_names)
+        self.channel_names = np.array(channel_names, dtype='S')
```
Comment (samuelgarcia) [inline neo/core/channelindex.py:181]: Please remove this. this is now fixed.

File: neo/io/basefromrawio.py [hunk 2]
```diff
@@ -139,7 +139,10 @@ def read_block(self, block_index=0, lazy=False, signal_group_mode=None,
                     chidx_annotations.pop('name')
                 chidx_annotations = check_annotations(chidx_annotations)
                 # this should be done with array_annotation soon:
-                ch_names = all_channels[ind_abs]['name'].astype('S')
+                ch_names = all_channels[ind_abs]['name'].astype('U')
+                # at this time ch_names are expected to be type 'S'.
+                # see https://github.com/NeuralEnsemble/python-neo/pull/824#issuecomment-644295081
+                ch_names = [name.encode('ascii', errors='replace') for name in ch_names]
                 neo_channel_index = ChannelIndex(index=ind_within,
```
Comment (samuelgarcia) [inline neo/io/basefromrawio.py:146]: Please remove this. this is now fixed.

**Comment by samuelgarcia** [comment]
@JuliaSprenger :Could we first merge this PR before  #775 which is also related to spike2io ?

**Comment by JuliaSprenger** [comment]
Sure, then we keep #775 in the queue.

**Comment by samuelgarcia** [comment]
In fact I ask exactly the contary I wanted.
I wanted to say:
"Could we merge first  #775 (mine older) and then #824 (this one)"

Sorry. One day I will read what I am writting before push the green button.

**Comment by MarinManuel** [comment]
@samuelgarcia Hi, I have reverted the commit on files channelindex.py and basefromrawio.py, is that good enough? Bit of a git newbie here...

**Comment by JuliaSprenger** [comment]
Hi @MarinManuel,
there are still conflicts between the current `master` branch on Neuralensemble and the master branch on your fork. What you can do is rebase your branch on top of the current Neuralensemble master. First make sure you also have an up-to-date local version of that one on your machine by running
`git fetch upstream` (assuming upstream points to Neuralensemble)
Then you can do the rebasing via
`git rebase upstream/master` - this will lead to the conflicts pointed out above to appear during the rebase  in your local files. The conflicts are indicated in the files as
```
<<<<
first set of changes
=======
second set of changes
>>>>>
```
Please edit the files to include all the changes that are suggested by Neuralensemble master and additionally keep the ones you want to include in this PR.
Finally you add the files using `git add <filename>` and run `git rebase --continue` to finalize the rebase process.
To update this PR you then push the changes you made to your fork: `git -f push origin master` (assuming origin points to your fork). Here you need the `-f` flag as you want to overwrite the history of the commits in your fork, which were changed during the rebase process.

I hope this helps. Tell us if there's issues during the process.

**Comment by MarinManuel** [comment]
Hi @JuliaSprenger 
Thank you for the detailed explanations. Hope this works now

**Comment by samuelgarcia** [comment]
Hi @MarinManuel 
could you modify theses two files like this:

neo/test/rawiotest/test_spike2rawio.py

```python
class TestSpike2RawIO(BaseTestRawIO, unittest.TestCase, ):
    rawioclass = Spike2RawIO
    files_to_download = [
        'File_spike2_1.smr',
        'File_spike2_2.smr',
        ...
        'Two-mice-bigfile-test000.smr', # <<<<<< new line
    ]
    entities_to_test = files_to_download
```

neo/test/iotest/test_spike2io.py
```python
class TestSpike2IO(BaseTestIO, unittest.TestCase, ):
    ioclass = Spike2IO
    files_to_test = [
        'File_spike2_1.smr',
        'File_spike2_2.smr',
        ...
        'Two-mice-bigfile-test000.smr', # <<<<<< new line
```
And afte rthis I will be able to merge this PR.

thanks again.

**Comment by MarinManuel** [comment]
> could you modify theses two files like this:
@samuelgarcia yes, done

**Comment by samuelgarcia** [comment]
thanks.
merging now.

**Pull request merged**